jni_bindings/src/lib.rs

```diff
@@ -457,273 +536,268 @@ pub extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1node
-            clone.throw_new(exception_class, message).unwrap();
-            clone.new_string("").unwrap().into()
-        }
+    let this = this as *mut Attribute;
+
+    if this.is_null() {
+        env.throw_new(
+            "java/lang/NullPointerException",
+            "Attribute::get_value called with null pointer",
+        )
+        .unwrap();
+
+        JObject::null().into()
+    } else {
+        let attr = &*this;
+        env.new_string(attr.value.to_str())
+            .unwrap_or_else(|_| JObject::null().into())
     }
 }
 
 #[no_mangle]
-pub extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1children<'local>(
+pub unsafe extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1children<
+    'local,
+>(
     mut env: JNIEnv<'local>,
     _: JClass<'local>,
+    // non-null raw pointer to Document
     this: jlong,
+    // NodeRef is the distinct u32 key mapped to a Node (and not a Node raw pointer)
     node_ref: jint,
 ) -> JIntArray<'local> {
-    let result = catch_unwind(|| {
-        // only u32 should be passed as node
-        let node = u32::try_from(node_ref).expect("value beyond `u32` range");
-        let node = NodeRef::new(node as usize);
-        let dom = unsafe { jlong_to_pointer::<dom::Document>(this).as_ref().unwrap() };
-        let children = dom.children(node);
-        let java_array = env.new_int_array(children.len() as jsize).unwrap();
+    let this = this as *mut Document;
+
+    if this.is_null() {
+        env.throw_new(
+            "java/lang/NullPointerException",
+            "Document::get_children called with null pointer",
+        )
+        .unwrap();
+
+        JObject::null().into()
+    } else {
+        let node = NodeRef::new(node_ref as usize);
+        let doc = &*this;
+        let children = doc.children(node);
+        let buff = env.new_int_array(children.len() as jsize).unwrap();
         let children: Vec<jint> = children.into_iter().map(|n| n.as_u32() as jint).collect();
-        env.set_int_array_region(&java_array, 0, &children).unwrap();
-        java_array
-    });
-
-    match result {
-        Ok(java_array) => java_array,
-        Err(_) => {
-            let exception_class = env.find_class("java/lang/RuntimeException").unwrap();
-            env.throw_new(exception_class, "Rust panic occurred")
-                .unwrap();
-            JObject::null().into()
-        }
+        env.set_int_array_region(&buff, 0, &children).unwrap();
+        buff
     }
 }
 
 // Java side should ensure only u32 is passed as the node parameter
 // Note! this function returns -1 when there's no parent
 #[no_mangle]
-pub extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1parent(
+pub unsafe extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1parent(
     mut env: JNIEnv,
     _: JClass,
+    // non-null raw pointer to Document
```

Comment:
The `non-null` can't be true or there would be no need for the `this.is_null()` check below.
```suggestion
    // raw pointer to Document
```

Reply:
The rationale for documenting it this way is that it's an exception for a null pointer to be passed. The null check is to manage this exception in a way that is sound and doesn't rely on catch_rewind